baby-liminal-extension/src/backend/mod.rs

```diff
@@ -78,12 +101,38 @@ where
     }
 
     /// Handle `verify` chain extension call.
-    pub fn verify<BackendExecutor: BackendExecutorT, Environment: EnvironmentT>(
+    pub fn verify<
+        BackendExecutor: BackendExecutorT,
+        Environment: EnvironmentT,
+        Weighting: WeightInfo,
+    >(
         mut env: Environment,
     ) -> ChainExtensionResult<RetVal> {
-        // todo: charge weight, validate args
+        // ------- Pre-charge optimistic weight. ---------------------------------------------------
+        let pre_charge = env.charge_weight(Weighting::verify())?;
+
+        // ------- Read the arguments. -------------------------------------------------------------
+        //
+        // TODO: charge additional weight for the args size (spam protection);
+        // this requires some benchmarking (maybe possible here, instead of polluting pallet's code)
+        // JIRA: https://cardinal-cryptography.atlassian.net/browse/A0-3578
         let args = env.read_as_unbounded(env.in_len())?;
-        let status = match BackendExecutor::verify(args) {
+
+        // ------- Forward the call. ---------------------------------------------------------------
+        let result = BackendExecutor::verify(args);
+
+        // ------- Adjust weight if needed. --------------------------------------------------------
+        match &result {
+            // In the failure case, if pallet provides us with a post-dispatch weight, we can make
+            // an adjustment.
+            Err((_, Some(actual_weight))) => env.adjust_weight(pre_charge, *actual_weight),
+            // Otherwise (positive case, or pallet doesn't provide us with any adjustment hint), we
+            // don't need to do anything.
+            Ok(_) | Err((_, None)) => {}
```

Comment:
This suggests that in the "Ok()" case we never need to adjust weight, which would suggest the `pre_charge` is super accurate. Not sure it's so easy -- but it surely depends on the details of the proof system. For a PoC we don't need to deal with these details.